### ISSUE TITLE
Section 4.1.4 Use Existing Credential "get"

### ISSUE THREAD

**Issue by rlin1** [w3c/webauthn#480]
in step 13.4.2 we say: "invoke the authenticatorGetAssertion operation on authenticator with rpId, clientDataHash, credentialList, and clientExtensions as parameters."

But this step is inside a "In parallel for each Credential C in credentialList".

I guess we wanted to say in 13.4.2:  "invoke the authenticatorGetAssertion operation on authenticator with rpId, clientDataHash, credential C, and clientExtensions as parameters."

Otherwise we would trigger the "authenticatorGetAssertion" with the *same* parameters multiple times (once for each entry in credentialList).

**Comment by rlin1** [comment]
Same is true for step 13.4.1

**Comment by equalsJeffH** [comment]
over in PR #427, @rlin1 said the below (https://github.com/w3c/webauthn/pull/427#issuecomment-306543869) which is relevant to this issue:

One more comment: the "in parallel for each credential C" seem to implicitly assume that each Authenticator related to the different credentials C could "display" (e.g. flash LED etc.) something to the use *in parallel*.
Note: If you have two bound authenticators in one smartphone (e.g. Fingerprint based and one Iris based one) this assumption might not be correct.

**Comment by rlin1** [comment]
Approach: 
1. create a list of all authenticators involved
2. do in parallel  for each authenticator in that list
    do the current step 1 and step 2 but the credentialList filtered by the specific authenticator.

**Comment by equalsJeffH** [comment]
It turns out that the above approach (https://github.com/w3c/webauthn/issues/480#issuecomment-308510804) is what is already in the spec at this time (and even before) (doh!). 

in looking further into this, I'm thinking we should perhaps try...

1. create a list of all authenticators involved
2. For each authenticator currently available on this platform, perform the following steps:
    1. If options.allowCredentials is not empty, find all creds listed in the latter that are bound to
       this authenticator. 
    2. If the latter list is not empty, then find all distinct transports listed in all authnr descriptors
     in the latter list. Let this list be distinctTransports.
    3. if the distinct transports list is not empty, then the client SHOULD select one transport 
    value from distinctTransports. Then, using transport, invoke in parallel the 
    authenticatorGetAssertion operation on authenticator, with rpId, clientDataHash, 
    allowCredentialDescriptorList, and authenticatorExtensions as parameters. 

[...]

..which, i think, would invoke the authnr once per collection of cred descriptors.

WDYT?

**Comment by rlin1** [comment]
Note that step 16 already does that, i.e. For each authenticator currently available on this platform, perform the following steps:.
As a consequence allowCredentialDescriptorList only contains the credentials bound to that specific authenticator (see step 16.2).

Additionally, I think the algorithm must distinguish 2 cases:
a) options.allowCredentialDescriptorList is non-empty
b) options.allowCredentialDescriptorList is empty

In case a) we might want to filter the number of authenticators connected to the platform (e.g. as used in step 16) to the ones related to one (or more) credentials in options.allowCredentialDescriptorList.

In case b) we should ask all connected authenticators whether they have a credential for the request RP.

One more issue: I think the transports are specific to the authenticator - not to the credential.  This means even if the platform has used BLE connection to talk to the authenticator while make(ing it), if the the platform known the authenticator is connected via USB at time of calling get, USB connection is the right transport here (for get).

--> Idea of an algo
replace current step 16 by
16a. if options.allowCredentialDescriptorList is empty then
     in parallel for each authenticator currently available on this platform, perform the following step:
         1. using local configuration knowledge of the appropriate transport to use with authenticator, invoke 
         the authenticatorGetAssertion operation on authenticator with rpId, clientDataHash, 
         empty allowCredentialDescriptorList, and clientExtensions as parameters.
16b. if options.allowCredentialDescriptorList is NOT empty then
    in parallel for each authenticator currently available on this platform, perform the following step:
       1. execute a platform-specific procedure to determine which, if any, public key credentials described by options.allowCredentials are bound to this authenticator, by matching with rpId, options.allowCredentials.id, and options.allowCredentials.type. Set allowCredentialDescriptorList to this filtered list.
       2. Determine the appropriate transport for talking to the authenticator by examining local platform knowledge and the transport field for each of the credentials in allowCredentialDescriptorList.
       3. Then, using transport, invoke the authenticatorGetAssertion operation on authenticator, with rpId, clientDataHash, allowCredentialDescriptorList, and authenticatorExtensions as parameters.

**Comment by equalsJeffH** [comment]
thanks much for your thoughts @rlin1 

my apologies, what I posted in https://github.com/w3c/webauthn/issues/480#issuecomment-308924747 is a summary abstraction of this portion of the alg, in a similar vein to the directly prior post https://github.com/w3c/webauthn/issues/480#issuecomment-308510804, and so omitted details. 

> Note that step 16 already does that, i.e. For each authenticator
> currently available on this platform, perform the following steps:.
> As a consequence allowCredentialDescriptorList only contains the
> credentials bound to that specific authenticator (see step 16.2).

understood. 

> Additionally, I think the algorithm must distinguish 2 cases: a)
> options.allowCredentialDescriptorList is non-empty b)
> options.allowCredentialDescriptorList is empty

agreed, the alg already does that in current step 16.3. I had essentially elided it in `[...]` (in https://github.com/w3c/webauthn/issues/480#issuecomment-308924747).

> In case a) we might want to filter the number of authenticators
> connected to the platform (e.g. as used in step 16) to the ones
> related to one (or more) credentials in `options.allowCredentialDescriptorList`.

Agreed -- the alg already does that in current step 16.2.   Also, `options.allowCredentialDescriptorList` should be `options.allowCredential` in the above.

> In case b) we should ask all [currently available] authenticators whether they
> have a credential [bound to] the request[ing] RP.

Agreed -- the alg already does that via current steps 15 & 16 & 16.3.

> One more issue: I think the transports are specific to the authenticator - not to the credential.  This means even if the platform has used BLE connection to talk to the authenticator while [creating the PK cred], if the the platform knows the authenticator is connected via USB at time of calling `get()`, USB connection is the right transport here (for `get()`).

yeah, I've also been thinking about that a bit.  Tho, I think it is something we should explore in a separate issue. 

```
> --> Idea of an algo
> replace current step 16 by
> 16a. if options.allowCredentialDescriptorList is empty then
```
You mean `options.allowCredentials`.  Also, this leaves out constructing `|allowCredentialDescriptorList|` (current step 16.2).

```
>      in parallel for each authenticator currently available on this 
>      platform, perform the following step:
>          1. using local configuration knowledge of the appropriate 
>             transport to use with authenticator, invoke 
>             the authenticatorGetAssertion operation on authenticator with rpId,
>              clientDataHash, [empty allowCredentialDescriptorList], 
>              and clientExtensions as parameters.
```
The current step 16 effectively does the above.  Note that we do not need to pass in `empty allowCredentialDescriptorList`, rather we just do not pass it in at all. 
```
> 16b. if options.allowCredentialDescriptorList is NOT empty then
```
You mean `options.allowCredentials`.
```
>     in parallel for each authenticator currently available on this platform, 
>     perform the following step:
>         1. execute a platform-specific procedure to determine which, if any,
>            public key credentials described by options.allowCredentials are
>            bound to this authenticator, by matching with rpId, 
>            options.allowCredentials.id, and options.allowCredentials.type. Set 
>             allowCredentialDescriptorList to this filtered list.
>         2. Determine the appropriate transport for talking to the
>            authenticator by examining local platform knowledge and the 
>            transport field for each of the credentials in
>            allowCredentialDescriptorList.
>         3. Then, using transport, invoke the authenticatorGetAssertion
>             operation on authenticator, with rpId, clientDataHash,
>            allowCredentialDescriptorList, and authenticatorExtensions as
>             parameters.
```

OK, this overall proposal ends up being essentially the same as the overall current step 16, except that:

a. it switches first on whether `options.allowCredentials` is empty or not (perhaps a worthwhile optimization?), and, 
b. it includes examining the cred descriptors' transport values. 

So, other than (a), this is what I was/am proposing.  

It seems we are largely on the same page. 

I have changes locally queued that essentially match this, except for (a).  I'll submit those as a PR. If we feel (a) is a worthwhile optimization, we can address it separately. OK?  thanks!

**Comment by rlin1** [comment]
If the caller doesn't provide any list of credentials, i.e. options.allowCredentials is empty, then all connected authenticators need to be asked.  This is the password replacement case.

Whether this is done as separate case a or implicitly is not that important.

**Comment by equalsJeffH** [comment]
@rlin1 
> If the caller doesn't provide any list of credentials, i.e. options.allowCredentials is empty, then all connected authenticators need to be asked. This is the password replacement case.
Whether this is done as separate case a or implicitly is not that important.

Ok, please review PR #495
